### PR TITLE
Human icon element that represents armor level and equipped module

### PR DESCRIPTION
**Pull request by XElectricX**
## About The Pull Request

Displays an icon over humans (like the health bar or job) to show how armored they are and what module they are using. There are 4 differently shaped icons: light, medium, heavy, uber. If a human has no armor or too little armor, no icon is displayed. Uber icon is exclusive to the tankiest of Marines (B18 or heavy with Tyr Mk2). Any mob can see this armor icon. In addition to the armor rating, the icon is overlayed with another one to display the module worn. If the player is under the threshold where an icon is not displayed, instead a generic box will be displayed as a background for the module icon.

The alternative could be a floating number under the player showing the exact percentage of damage reduction but that would be too technical and go against the fun of it. Might look ugly too. An armor bar like plasma or health could work but that takes up a lot of space compared to other mob icons. This should be a nice compromise unless someone has some better ideas.

### Legend
image

## Why It's Good For The Game

A player should always be able to discern another's equipment. Visual clarity has been neglected in favor of style. It is unlikely we can ever go back to knowing what armor a player is wearing at glance since there are so many armor sets, and so many different armor pieces on top of that. Players in 2018 only had to recognize 4 different armor sets and there were no capes to hide under. Modules also exist now, so it's important to let players be aware of that too.

## Changelog
:cl:
add: Icon over human players that represents their armor level and equipped module. Can be seen by all players.
del: Remove green Vali icon over the player.
imageadd: 4 different shield icons and an icon for each module.
/:cl:

### CODE REVIEW

**Comment by Lumipharon** [comment]
More clogged up UI that can't be turned off is :(
Also well as only reflecting melee armour.

**Comment by Victor239** [comment]
> only reflecting melee armour.

This could be a UI element only available on XvH then. Then again it's also good to know stuff like Vali and armour level on HvH also.

**Comment by dopamiin0** [comment]
thank god. its very annoying as xeno trying to play "guess how tanky the marine is" based off of an extremely unclear visual and their move speed.

**Comment by Victor239** [comment]
Some of the icons are quite similar:
- Mimir and Vali
- Hod and Baldur
- Valkyrie and Svalinn

They could use with their icon differentiated more, and perhaps consider using different colours like purple, orange, black, brown, pink, etc.

File: code/controllers/configuration/entries/game_options.dm [hunk 1]
```diff
@@ -121,3 +121,6 @@
 	min_val = 1
 
 /datum/config_entry/str_list/tts_voice_blacklist
+
+/datum/config_entry/flag/armor_hud
+	config_entry_value = TRUE
```
Comment (TiviPlus) [inline code/controllers/configuration/entries/game_options.dm:126]: Remove the config please

File: code/datums/components/chem_booster.dm [hunk 2]
```diff
@@ -384,9 +389,15 @@
 /datum/component/chem_booster/proc/update_resource(amount)
 	var/amount_added = min(resource_storage_max - resource_storage_current, amount)
 	resource_storage_current = max(resource_storage_current + amount_added, 0)
-	wearer.overlays -= resource_overlay
-	resource_overlay.alpha = resource_storage_current/resource_storage_max*255
-	wearer.overlays += resource_overlay
+	//Check if reagent level was empty and now is not, or if it was not empty but now is
+	if(CONFIG_GET(flag/armor_hud) && ishuman(wearer) && ((!resource_storage_current && amount_added) || resource_storage_current == amount_added))
+		hud_resource_update()
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:394]: dont you already have a signal for this?

File: code/game/data_huds.dm [hunk 3]
```diff
@@ -625,6 +625,87 @@
...
+			if(/obj/item/armor_module/module/mimir_environment_protection,
+			   /obj/item/armor_module/module/mimir_environment_protection/mark1,
+			   /obj/item/armor_module/module/mimir_environment_protection/som)
+				holder.overlays += image(holder.icon, src, "biohazard")
+			if(/obj/item/armor_module/module/tyr_extra_armor,
+			   /obj/item/armor_module/module/tyr_extra_armor/mark1,
+			   /obj/item/armor_module/module/tyr_extra_armor/som)
+				holder.overlays += image(holder.icon, src, "armor")
+			if(/obj/item/armor_module/module/valkyrie_autodoc, /obj/item/armor_module/module/valkyrie_autodoc/som)
+				holder.overlays += image(holder.icon, src, "medic")
```
Comment (TiviPlus) [inline code/game/data_huds.dm:675]: use a var on the module 

eg.

/valk
  hud_icon = "medic"

File: code/datums/components/chem_booster.dm (hunk 2 above)
Comment (XElectricX) [reply code/datums/components/chem_booster.dm:394]: The signal is only sent when equipping armor with the Vali module. Needed for the first time icon display.

File: code/game/data_huds.dm (hunk 3 above)
Comment (XElectricX) [reply code/game/data_huds.dm:675]: Good idea.

File: code/game/data_huds.dm [hunk 4]
```diff
@@ -625,6 +625,68 @@
...
+		/*
+		I COULD have it do overlays += module.icon_state but if the icon_state is renamed, it would silently break
+		This way if someone renames a module's sprite, it will still work
+		The hud_icon var just has to be set whenever a new module is added
+		*/
+		if(istype(module, /obj/item/armor_module/module/chemsystem))
+			//chem_booster/proc/hud_update_resource() handles what sprite to assign
+			SEND_SIGNAL(src, COMSIG_CHEMSYSTEM_RESOURCE_UPDATE)
+		else
+			holder.overlays += image(holder.icon, src, module.hud_icon)
```
Comment (TiviPlus) [inline code/game/data_huds.dm:653]: I dunno about havinf this a signal this seems excessively convoluted (signal >component >human proc)

Rest lgtm

Comment (XElectricX) [reply code/game/data_huds.dm:653]: The fault of the chem stuff being a component. I already hammered my head for a few hours trying to think of an alternative way but with how components are encapsulated...

**Review by Kuro020 — CHANGES_REQUESTED** [review]
These need to be smaller

Me personally I wouldnt even merge this pr, last I checked marines cant see xeno armor level and sunder and have to play a guessing game so idk why xenos should have this, no codex doesnt count, no soul in this game reads the codex, not even the people adding stuff to the codex

**Comment by Joe13413** [comment]
> These need to be smaller
> 
> Me personally I wouldnt even merge this pr, last I checked marines cant see xeno armor level and sunder and have to play a guessing game so idk why xenos should have this, no codex doesnt count, no soul in this game reads the codex, not even the people adding stuff to the codex

dude uhhhh knowing marine armor levels would add some strategy to who you pick to fuck with as xeno
and every marine knows every xenos armor level by DEFAULT due to wiki listing all armor values

If i was a rouny, and i saw a mfer with heavy armor + tyr2 beside a bro with light armor, i'm picking the light armor. The thing is, i can't tell what they're running because of all the capes + different armor sprites + all that jazz

you don't need to guess what armor xenos have when you're a marine, you already know what they've got. You absolutely do need to guess as a xeno

**Comment by KillerOrcaCora** [comment]
This for the most part looks terrible in-game, even with the sprites trimmed.

**Comment by Joe13413** [comment]
> Says who, who knows xeno armor levels? people have vague ideas of it, unless you code dive you're never gonna find out, also I could literally say the same for "well if xenos had armor level indicators I would know when to pull my buckshot instead of using my rifle to do more damage" but thats one ticket to busy visuals hell.

its called a wiki you slackjaw

**Pull request closed without merging**